// @patternfly/react-table/dist/esm/components/Table/base/provider.js
function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/**
 * provider.js
 *
 * Forked from reactabular-table version 8.14.0
 * https://github.com/reactabular/reactabular/tree/v8.14.0/packages/reactabular-table/src
 * */
import React from 'react';
import PropTypes from 'prop-types';
import { tableTypes, tableDefaults, tableContextTypes } from './types';
const componentDefaults = tableDefaults.renderers;
export default class Provider extends React.Component {
  getChildContext() {
    const {
      columns,
      components,
      renderers
    } = this.props;
    let finalRenderers = renderers; // XXXXX: Drop in the next major version

    if (components) {
      // eslint-disable-next-line no-console
      console.warn('`components` have been deprecated in favor of `renderers` and will be removed in the next major version, please rename!');
      finalRenderers = components;
    }

    return {
      columns,
      renderers: {
        table: finalRenderers.table || componentDefaults.table,
        header: _objectSpread({}, componentDefaults.header, finalRenderers.header),
        body: _objectSpread({}, componentDefaults.body, finalRenderers.body)
      }
    };
  }

  render() {
    const _this$props = this.props,
          {
      columns,
      // eslint-disable-line no-unused-vars
      renderers,
      components,
      // XXXXX: Drop in the next major version
      children
    } = _this$props,
          props = _objectWithoutProperties(_this$props, ["columns", "renderers", "components", "children"]);

    return React.createElement(renderers.table || tableDefaults.renderers.table, props, children);
  }

}
Provider.propTypes = _objectSpread({}, tableTypes, {
  children: PropTypes.any
});
Provider.defaultProps = _objectSpread({}, tableDefaults);
Provider.childContextTypes = tableContextTypes;
