// @patternfly/react-table/dist/js/components/Table/base/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tableDefaults = exports.tableHeaderRowDefaults = exports.tableHeaderRowTypes = exports.tableHeaderContextTypes = exports.tableHeaderTypes = exports.tableBodyRowDefaults = exports.tableBodyRowTypes = exports.tableBodyContextTypes = exports.tableBodyDefaults = exports.tableBodyTypes = exports.tableContextTypes = exports.tableTypes = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 * types.js
 *
 * Forked from reactabular-table version 8.14.0
 * https://github.com/reactabular/reactabular/tree/v8.14.0/packages/reactabular-table/src
 * */
var arrayOfObjectColumns = _propTypes["default"].arrayOf(_propTypes["default"].shape({
  header: _propTypes["default"].shape({
    label: _propTypes["default"].string,
    transforms: _propTypes["default"].arrayOf(_propTypes["default"].func),
    formatters: _propTypes["default"].arrayOf(_propTypes["default"].func),
    props: _propTypes["default"].object
  }),
  cell: _propTypes["default"].shape({
    property: _propTypes["default"].oneOfType([_propTypes["default"].number, _propTypes["default"].string]),
    transforms: _propTypes["default"].arrayOf(_propTypes["default"].func),
    formatters: _propTypes["default"].arrayOf(_propTypes["default"].func),
    props: _propTypes["default"].object
  })
}));

var arrayOfArrayColumns = _propTypes["default"].arrayOf(_propTypes["default"].array);

var rowsType = _propTypes["default"].oneOfType([arrayOfObjectColumns, arrayOfArrayColumns]);

var rowKeyType = _propTypes["default"].oneOfType([_propTypes["default"].func, _propTypes["default"].string]);

var rowDataType = _propTypes["default"].oneOfType([_propTypes["default"].array, _propTypes["default"].object]);

var tableTypes = {
  columns: _propTypes["default"].array.isRequired,
  renderers: _propTypes["default"].object,
  components: _propTypes["default"].object // XXXXX: Deprecated in favor of renderers, remove in the next major!

};
exports.tableTypes = tableTypes;
var tableContextTypes = {
  columns: _propTypes["default"].array.isRequired,
  renderers: _propTypes["default"].object
};
exports.tableContextTypes = tableContextTypes;
var tableBodyDefaults = {
  onRow: function onRow() {}
};
exports.tableBodyDefaults = tableBodyDefaults;
var tableBodyTypes = {
  onRow: _propTypes["default"].func,
  rows: rowsType.isRequired,
  rowKey: rowKeyType
};
exports.tableBodyTypes = tableBodyTypes;
var tableBodyContextTypes = {
  columns: _propTypes["default"].array.isRequired,
  renderers: _propTypes["default"].object
};
exports.tableBodyContextTypes = tableBodyContextTypes;
var tableBodyRowDefaults = {
  onRow: function onRow() {
    return {};
  }
};
exports.tableBodyRowDefaults = tableBodyRowDefaults;
var tableBodyRowTypes = {
  columns: _propTypes["default"].array.isRequired,
  renderers: _propTypes["default"].object,
  onRow: _propTypes["default"].func,
  rowIndex: _propTypes["default"].number.isRequired,
  rowData: rowDataType.isRequired,
  rowKey: _propTypes["default"].string.isRequired
};
exports.tableBodyRowTypes = tableBodyRowTypes;
var tableHeaderTypes = {
  headerRows: _propTypes["default"].arrayOf(arrayOfObjectColumns),
  children: _propTypes["default"].any
};
exports.tableHeaderTypes = tableHeaderTypes;
var tableHeaderContextTypes = {
  columns: _propTypes["default"].array.isRequired,
  renderers: _propTypes["default"].object
};
exports.tableHeaderContextTypes = tableHeaderContextTypes;
var tableHeaderRowDefaults = {
  onRow: function onRow() {
    return {};
  }
};
exports.tableHeaderRowDefaults = tableHeaderRowDefaults;
var tableHeaderRowTypes = {
  renderers: _propTypes["default"].object,
  onRow: _propTypes["default"].func,
  rowIndex: _propTypes["default"].number.isRequired,
  rowData: rowDataType.isRequired
};
exports.tableHeaderRowTypes = tableHeaderRowTypes;
var tableDefaults = {
  renderers: {
    table: 'table',
    header: {
      wrapper: 'thead',
      row: 'tr',
      cell: 'th'
    },
    body: {
      wrapper: 'tbody',
      row: 'tr',
      cell: 'td'
    }
  }
};
exports.tableDefaults = tableDefaults;
