// @patternfly/react-table/dist/umd/test-helpers/data-sets.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.actions = exports.rows = exports.columns = undefined;

  var _react2 = _interopRequireDefault(_react);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  /* eslint-disable no-console */
  const columns = exports.columns = [{
    title: 'Header cell'
  }, 'Branches', {
    title: 'Pull requests'
  }, 'Workspaces', {
    title: 'Last Commit'
  }];
  const rows = exports.rows = [{
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }, {
    cells: ['one', 'two', 'three', 'four', 'five']
  }];
  const actions = exports.actions = [{
    title: 'Some action',
    onClick: (event, rowId) => console.log('clicked on Some action, on row: ', rowId)
  }, {
    title: _react2.default.createElement("div", null, "Another action"),
    onClick: (event, rowId) => console.log('clicked on Another action, on row: ', rowId)
  }, {
    isSeparator: true
  }, {
    title: 'Third action',
    onClick: (event, rowId) => console.log('clicked on Third action, on row: ', rowId)
  }];
});
