// @patternfly/react-core/dist/esm/layouts/Grid/GridItem.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from 'react';
import PropTypes from 'prop-types';
import styles from '@patternfly/react-styles/css/layouts/Grid/grid';
import { css } from '@patternfly/react-styles';
import { DeviceSizes } from '../../styles/sizes';
import { getOffsetKey, getOffsetModifier, getRowSpanKey, getRowSpanModifier, getSpanModifier, gridSpans } from './gridUtils';
const propTypes = {
  /** content rendered inside the Grid Layout Item */
  children: PropTypes.any,

  /** additional classes added to the Grid Layout Item */
  className: PropTypes.string,

  /** the number of columns the grid item spans. Value should be a number 1-12   */
  span: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans. Value should be a number 1-12   */
  rowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns a grid item is offset */
  offset: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item spans on small device. Value should be a number 1-12   */
  sm: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans on medium device. Value should be a number 1-12   */
  smRowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item is offset on small device. Value should be a number 1-12   */
  smOffset: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item spans on medium device. Value should be a number 1-12   */
  md: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans on medium device. Value should be a number 1-12   */
  mdRowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item is offset on medium device. Value should be a number 1-12   */
  mdOffset: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item spans on large device. Value should be a number 1-12   */
  lg: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans on large device. Value should be a number 1-12   */
  lgRowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item is offset on large device. Value should be a number 1-12   */
  lgOffset: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item spans on xLarge device. Value should be a number 1-12   */
  xl: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans on large device. Value should be a number 1-12   */
  xlRowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item is offset on xLarge device. Value should be a number 1-12   */
  xlOffset: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item spans on 2xLarge device. Value should be a number 1-12   */
  xl2: PropTypes.oneOf(gridSpans),

  /** the number of rows the grid item spans on 2xLarge device. Value should be a number 1-12   */
  xl2RowSpan: PropTypes.oneOf(gridSpans),

  /** the number of columns the grid item is offset on 2xLarge device. Value should be a number 1-12   */
  xl2Offset: PropTypes.oneOf(gridSpans),

  /** Additional props are spread to the container <div> */
  '': PropTypes.any // eslint-disable-line react/require-default-props

};
const defaultProps = {
  children: null,
  className: '',
  span: null,
  rowSpan: null,
  offset: null,
  sm: null,
  smRowSpan: null,
  smOffset: null,
  md: null,
  mdRowSpan: null,
  mdOffset: null,
  lg: null,
  lgRowSpan: null,
  lgOffset: null,
  xl: null,
  xlRowSpan: null,
  xlOffset: null,
  xl2: null,
  xl2RowSpan: null,
  xl2Offset: null
};

const GridItem = (_ref) => {
  let {
    children,
    className,
    span,
    rowSpan,
    offset
  } = _ref,
      props = _objectWithoutProperties(_ref, ["children", "className", "span", "rowSpan", "offset"]);

  const classes = [styles.gridItem, span && getSpanModifier(span), offset && getOffsetModifier(offset), rowSpan && getRowSpanModifier(rowSpan)];
  Object.entries(DeviceSizes).forEach(([propKey, classModifier]) => {
    const spanValue = props[propKey];
    const rowSpanValue = props[getRowSpanKey(propKey)];
    const offsetValue = props[getOffsetKey(propKey)];

    if (spanValue) {
      classes.push(getSpanModifier(spanValue, classModifier));
    }

    if (rowSpanValue) {
      classes.push(getRowSpanModifier(rowSpanValue, classModifier));
    }

    if (offsetValue) {
      classes.push(getOffsetModifier(offsetValue, classModifier));
    }

    delete props[propKey];
    delete props[getRowSpanKey(propKey)];
    delete props[getOffsetKey(propKey)];
  });
  return React.createElement("div", _extends({
    className: css(...classes, className)
  }, props), children);
};

GridItem.propTypes = propTypes;
GridItem.defaultProps = defaultProps;
export default GridItem;
