// @patternfly/react-core/dist/esm/components/AboutModal/AboutModal.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { css } from '@patternfly/react-styles';
import styles from '@patternfly/react-styles/css/components/Backdrop/backdrop';
import { canUseDOM } from 'exenv';
import { KEY_CODES } from '../../helpers/constants';
import { AboutModalContainer } from './AboutModalContainer';
;
export class AboutModal extends React.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "id", AboutModal.currentId++);

    _defineProperty(this, "ariaLabelledBy", `pf-about-modal-title-${this.id}`);

    _defineProperty(this, "ariaDescribedBy", `pf-about-modal-content-${this.id}`);

    _defineProperty(this, "handleEscKeyClick", event => {
      if (event.keyCode === KEY_CODES.ESCAPE_KEY && this.props.isOpen) {
        this.props.onClose();
      }
    });

    _defineProperty(this, "appendContainer", () => {
      if (!this.container) {
        this.container = document.createElement('div');
        document.body.appendChild(this.container);
        document.addEventListener('keydown', this.handleEscKeyClick, false);
      }

      if (this.props.isOpen) {
        document.body.classList.add(css(styles.backdropOpen));
      } else {
        document.body.classList.remove(css(styles.backdropOpen));
      }
    });

    if (props.brandImageSrc && !props.brandImageAlt) {
      // tslint:disable-next-line:no-console
      console.error('AboutModal:', 'brandImageAlt is required when a brandImageSrc is specified');
    }
  }

  componentDidMount() {
    this.appendContainer();
  }

  componentDidUpdate() {
    if (this.props.isOpen) {
      document.body.classList.add(css(styles.backdropOpen));
    } else {
      document.body.classList.remove(css(styles.backdropOpen));
    }
  }

  componentWillUnmount() {
    if (this.container) {
      document.body.removeChild(this.container);
    }

    document.removeEventListener('keydown', this.handleEscKeyClick, false);
  }

  render() {
    if (!canUseDOM) {
      return null;
    }

    if (canUseDOM && !this.container) {
      this.appendContainer();
    }

    return ReactDOM.createPortal(React.createElement(AboutModalContainer, _extends({
      ariaLabelledbyId: this.ariaLabelledBy,
      ariaDescribedById: this.ariaDescribedBy
    }, this.props)), this.container);
  }

}

_defineProperty(AboutModal, "propTypes", {
  children: _pt.node.isRequired,
  className: _pt.string,
  isOpen: _pt.bool,
  onClose: _pt.func,
  productName: _pt.string,
  trademark: _pt.string,
  brandImageSrc: _pt.string.isRequired,
  brandImageAlt: _pt.string.isRequired,
  backgroundImageSrc: _pt.string,
  noAboutModalBoxContentContainer: _pt.bool
});

_defineProperty(AboutModal, "currentId", 0);

_defineProperty(AboutModal, "defaultProps", {
  className: '',
  isOpen: false,
  onClose: () => undefined,
  productName: '',
  trademark: '',
  backgroundImageSrc: '',
  noAboutModalBoxContentContainer: false
});
