// @patternfly/react-core/dist/esm/components/Popover/Popover.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import Tippy from '@tippy.js/react';
import { KEY_CODES } from '../../helpers/constants';
import styles from '@patternfly/react-styles/css/components/Popover/popover';
import { css, getModifier } from '@patternfly/react-styles';
import { PopoverArrow } from './PopoverArrow';
import { PopoverContent } from './PopoverContent';
import { PopoverBody } from './PopoverBody';
import { PopoverHeader } from './PopoverHeader';
import { PopoverFooter } from './PopoverFooter';
import { PopoverCloseButton } from './PopoverCloseButton';
import GenerateId from '../../helpers/GenerateId/GenerateId';
import { c_popover_MaxWidth as popoverMaxWidth } from '@patternfly/react-tokens';
import { tippyStyles } from '../Tooltip/styles';

// Can't use ES6 imports :(
// The types for it are also wrong, we should probably ditch this dependency.
// tslint:disable-next-line
const FocusTrap = require('focus-trap-react');

tippyStyles();
export let PopoverPosition;

(function (PopoverPosition) {
  PopoverPosition["top"] = "top";
  PopoverPosition["bottom"] = "bottom";
  PopoverPosition["left"] = "left";
  PopoverPosition["right"] = "right";
})(PopoverPosition || (PopoverPosition = {}));

;
export class Popover extends React.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "hideOrNotify", () => {
      if (this.props.isVisible === null) {
        // Handle closing
        this.tip.hide();
      } else {
        // notify consumer
        this.props.shouldClose(this.tip);
      }
    });

    _defineProperty(this, "handleEscKeyClick", event => {
      if (event.keyCode === KEY_CODES.ESCAPE_KEY && this.tip.state.isVisible) {
        this.hideOrNotify();
      }
    });

    _defineProperty(this, "storeTippyInstance", tip => {
      this.tip = tip;
    });

    _defineProperty(this, "closePopover", () => {
      this.hideOrNotify();
    });

    _defineProperty(this, "hideAllPopovers", () => {
      document.querySelectorAll('.tippy-popper').forEach(popper => {
        if (popper._tippy) {
          popper._tippy.hide();
        }
      });
    });

    _defineProperty(this, "onHide", tip => {
      if (this.state.isOpen) {
        this.setState({
          isOpen: false
        });
      }

      return this.props.onHide(tip);
    });

    _defineProperty(this, "onHidden", tip => this.props.onHidden(tip));

    _defineProperty(this, "onMount", tip => this.props.onMount(tip));

    _defineProperty(this, "onShow", tip => {
      const {
        hideOnOutsideClick,
        isVisible,
        onShow
      } = this.props; // hide all other open popovers first if events are managed by us

      if (!hideOnOutsideClick && isVisible === null) {
        this.hideAllPopovers();
      }

      if (this.state.isOpen === false) {
        this.setState({
          isOpen: true
        });
      }

      return onShow(tip);
    });

    _defineProperty(this, "onShown", tip => this.props.onShown(tip));

    this.state = {
      isOpen: false
    };
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleEscKeyClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleEscKeyClick, false);
  }

  render() {
    const _this$props = this.props,
          {
      position,
      enableFlip,
      children,
      className,
      'aria-label': ariaLabel,
      headerContent,
      bodyContent,
      footerContent,
      isVisible,
      shouldClose,
      appendTo,
      hideOnOutsideClick,
      onHide,
      onHidden,
      onShow,
      onShown,
      onMount,
      zIndex,
      maxWidth,
      closeBtnAriaLabel
    } = _this$props,
          rest = _objectWithoutProperties(_this$props, ["position", "enableFlip", "children", "className", "aria-label", "headerContent", "bodyContent", "footerContent", "isVisible", "shouldClose", "appendTo", "hideOnOutsideClick", "onHide", "onHidden", "onShow", "onShown", "onMount", "zIndex", "maxWidth", "closeBtnAriaLabel"]);

    if (!headerContent && !ariaLabel) {
      return new Error('aria-label is required when header is not used');
    }

    const content = React.createElement(GenerateId, null, randomId => this.state.isOpen && React.createElement(FocusTrap, {
      focusTrapOptions: {
        clickOutsideDeactivates: true
      }
    }, React.createElement("div", _extends({
      className: css(styles.popover, !enableFlip && getModifier(styles, position, styles.modifiers.top), className),
      role: "dialog",
      "aria-modal": "true",
      "aria-label": headerContent ? undefined : ariaLabel,
      "aria-labelledby": headerContent ? `popover-${randomId}-header` : undefined,
      "aria-describedby": `popover-${randomId}-body`
    }, rest), React.createElement(PopoverArrow, null), React.createElement(PopoverContent, null, React.createElement(PopoverCloseButton, {
      onClose: this.closePopover,
      "aria-label": closeBtnAriaLabel
    }), headerContent && React.createElement(PopoverHeader, {
      id: `popover-${randomId}-header`
    }, headerContent), React.createElement(PopoverBody, {
      id: `popover-${randomId}-body`
    }, bodyContent), footerContent && React.createElement(PopoverFooter, null, footerContent)))));
    const handleEvents = isVisible === null;

    const shouldHideOnClick = () => {
      if (handleEvents) {
        if (hideOnOutsideClick === true) {
          return true;
        }

        return 'toggle';
      }

      return false;
    };

    return React.createElement(Tippy, {
      onCreate: this.storeTippyInstance,
      maxWidth: maxWidth,
      zIndex: zIndex,
      appendTo: appendTo,
      content: content,
      lazy: true,
      trigger: handleEvents ? 'click' : 'manual',
      isVisible: isVisible,
      hideOnClick: shouldHideOnClick(),
      animateFill: false,
      theme: "pf-tippy",
      performance: true,
      interactive: true,
      interactiveBorder: 0,
      placement: position,
      distance: 25,
      flip: enableFlip,
      popperOptions: {
        modifiers: {
          preventOverflow: {
            enabled: enableFlip
          },
          hide: {
            enabled: enableFlip
          }
        }
      },
      onHide: this.onHide,
      onHidden: this.onHidden,
      onShow: this.onShow,
      onShown: this.onShown,
      onMount: this.onMount
    }, children);
  }

}

_defineProperty(Popover, "propTypes", {
  position: _pt.oneOf(['top', 'bottom', 'left', 'right']),
  enableFlip: _pt.bool,
  className: _pt.string,
  children: _pt.element.isRequired,
  'aria-label': _pt.string,
  headerContent: _pt.node,
  bodyContent: _pt.node.isRequired,
  footerContent: _pt.node,
  isVisible: _pt.bool,
  shouldClose: _pt.func,
  appendTo: _pt.oneOfType([_pt.element, _pt.func]),
  hideOnOutsideClick: _pt.bool,
  onHide: _pt.func,
  onHidden: _pt.func,
  onShow: _pt.func,
  onShown: _pt.func,
  onMount: _pt.func,
  zIndex: _pt.number,
  maxWidth: _pt.string,
  closeBtnAriaLabel: _pt.string
});

_defineProperty(Popover, "defaultProps", {
  position: 'top',
  enableFlip: true,
  className: '',
  isVisible: null,
  shouldClose: () => null,
  'aria-label': '',
  headerContent: null,
  footerContent: null,
  appendTo: () => document.body,
  hideOnOutsideClick: true,
  onHide: () => null,
  onHidden: () => null,
  onShow: () => null,
  onShown: () => null,
  onMount: () => null,
  zIndex: 9999,
  maxWidth: popoverMaxWidth && popoverMaxWidth.value,
  closeBtnAriaLabel: 'Close'
});
