// @patternfly/react-table/dist/umd/components/Table/ActionsColumn.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "@patternfly/react-core"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("@patternfly/react-core"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.reactCore);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _reactCore) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  const propTypes = {
    children: _propTypes2.default.node,
    items: _propTypes2.default.array,
    isDisabled: _propTypes2.default.bool,
    dropdownPosition: _propTypes2.default.oneOf(Object.values(_reactCore.DropdownPosition)),
    dropdownDirection: _propTypes2.default.oneOf(Object.values(_reactCore.DropdownDirection)),
    rowData: _propTypes2.default.object,
    extraData: _propTypes2.default.shape({
      rowIndex: _propTypes2.default.number,
      columnIndex: _propTypes2.default.number,
      column: _propTypes2.default.object,
      property: _propTypes2.default.string
    })
  };
  const defaultProps = {
    children: null,
    items: [],
    isDisabled: false,
    dropdownPosition: _reactCore.DropdownPosition.right,
    dropdownDirection: _reactCore.DropdownDirection.down,
    rowData: {},
    extraData: {}
  };

  class ActionsColumn extends _react2.default.Component {
    constructor(props) {
      super(props);

      _defineProperty(this, "onToggle", isOpen => {
        this.setState({
          isOpen
        });
      });

      _defineProperty(this, "onSelect", (event, onClick) => {
        const {
          rowData,
          extraData
        } = this.props;
        event.preventDefault();
        onClick && onClick(event, extraData.rowIndex, rowData, extraData);
        this.setState({
          isOpen: !this.state.isOpen
        });
      });

      this.state = {
        isOpen: false
      };
    }

    render() {
      const {
        isOpen
      } = this.state;
      const {
        items,
        children,
        dropdownPosition,
        dropdownDirection,
        isDisabled
      } = this.props;
      return _react2.default.createElement(_react2.default.Fragment, null, _react2.default.createElement(_reactCore.Dropdown, {
        onToggle: this.onToggle,
        toggle: _react2.default.createElement(_reactCore.KebabToggle, {
          isDisabled: isDisabled,
          onToggle: this.onToggle
        }),
        position: dropdownPosition,
        direction: dropdownDirection,
        isOpen: isOpen,
        dropdownItems: items.map((_ref, key) => {
          let {
            title,
            itemKey,
            onClick,
            isSeparator
          } = _ref,
              props = _objectWithoutProperties(_ref, ["title", "itemKey", "onClick", "isSeparator"]);

          return isSeparator ? _react2.default.createElement(_reactCore.DropdownSeparator, _extends({}, props, {
            key: itemKey || key,
            "data-key": itemKey || key
          })) : _react2.default.createElement(_reactCore.DropdownItem, _extends({
            onClick: event => this.onSelect(event, onClick)
          }, props, {
            key: itemKey || key,
            "data-key": itemKey || key
          }), title);
        }),
        isPlain: true
      }), children);
    }

  }

  ActionsColumn.propTypes = propTypes;
  ActionsColumn.defaultProps = defaultProps;
  exports.default = ActionsColumn;
});
