// @patternfly/react-table/dist/umd/components/Table/BodyWrapper.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "@patternfly/react-styles/css/components/Table/table", "@patternfly/react-styles", "./utils/headerUtils"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("@patternfly/react-styles/css/components/Table/table"), require("@patternfly/react-styles"), require("./utils/headerUtils"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.table, global.reactStyles, global.headerUtils);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _table, _reactStyles, _headerUtils) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _table2 = _interopRequireDefault(_table);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  // eslint-disable-next-line react/prefer-stateless-function
  class BodyWrapper extends _react2.default.Component {
    render() {
      const _this$props = this.props,
            {
        mappedRows,
        tbodyRef,
        rows,
        onCollapse
      } = _this$props,
            props = _objectWithoutProperties(_this$props, ["mappedRows", "tbodyRef", "rows", "onCollapse"]);

      if (mappedRows && mappedRows.some(row => row.hasOwnProperty('parent'))) {
        return _react2.default.createElement(_react2.default.Fragment, null, (0, _headerUtils.mapOpenedRows)(mappedRows, this.props.children).map((oneRow, key) => _react2.default.createElement("tbody", _extends({}, props, {
          className: (0, _reactStyles.css)(oneRow.isOpen && _table2.default.modifiers.expanded),
          key: `tbody-${key}`,
          ref: tbodyRef
        }), oneRow.rows)));
      }

      return _react2.default.createElement("tbody", _extends({}, props, {
        ref: tbodyRef
      }));
    }

  }

  BodyWrapper.propTypes = {
    children: _propTypes2.default.node,
    mappedRows: _propTypes2.default.array,
    rows: _propTypes2.default.array,
    onCollapse: _propTypes2.default.func,
    tbodyRef: _propTypes2.default.func
  };
  BodyWrapper.defaultProps = {
    children: null,
    mappedRows: undefined,
    rows: [],
    onCollapse: null,
    tbodyRef: null
  };
  exports.default = BodyWrapper;
});
