// @patternfly/react-core/dist/umd/components/Dropdown/dropdownConstants.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.DropdownArrowContext = exports.DropdownContext = exports.DropdownDirection = exports.DropdownPosition = undefined;

  var _react2 = _interopRequireDefault(_react);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const DropdownPosition = exports.DropdownPosition = {
    right: 'right',
    left: 'left'
  };
  const DropdownDirection = exports.DropdownDirection = {
    up: 'up',
    down: 'down'
  };

  const DropdownContext = exports.DropdownContext = _react2.default.createContext({
    onSelect: () => {},
    menuClass: '',
    itemClass: '',
    toggleClass: '',
    baseClass: '',
    baseComponent: 'div',
    sectionClass: '',
    sectionTitleClass: '',
    sectionComponent: 'section',
    disabledClass: '',
    hoverClass: '',
    separatorClass: ''
  });

  const DropdownArrowContext = exports.DropdownArrowContext = _react2.default.createContext({
    keyHandler: null,
    sendRef: null
  });
});
