// @patternfly/react-core/dist/umd/components/Dropdown/DropdownMenu.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "@patternfly/react-styles/css/components/Dropdown/dropdown", "@patternfly/react-styles", "../../helpers/componentShape", "react-dom", "../../helpers/util", "./dropdownConstants", "../../helpers/constants"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("@patternfly/react-styles/css/components/Dropdown/dropdown"), require("@patternfly/react-styles"), require("../../helpers/componentShape"), require("react-dom"), require("../../helpers/util"), require("./dropdownConstants"), require("../../helpers/constants"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.dropdown, global.reactStyles, global.componentShape, global.reactDom, global.util, global.dropdownConstants, global.constants);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _dropdown, _reactStyles, _componentShape, _reactDom, _util, _dropdownConstants, _constants) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _dropdown2 = _interopRequireDefault(_dropdown);

  var _reactDom2 = _interopRequireDefault(_reactDom);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  const propTypes = {
    /** Anything which can be rendered as dropdown items */
    children: _propTypes2.default.node,

    /** Classess applied to root element of dropdown menu */
    className: _propTypes2.default.string,

    /** Flag to indicate if menu is opened */
    isOpen: _propTypes2.default.bool,

    /** Flag to indicate if menu should be opened on enter */
    openedOnEnter: _propTypes2.default.bool,

    /** Indicates which component will be used as dropdown menu */
    component: _componentShape.componentShape,

    /** Indicates where menu will be alligned horizontally */
    position: _propTypes2.default.oneOf(Object.values(_dropdownConstants.DropdownPosition)),

    /** Flag to indicate if menu is grouped */
    isGrouped: _propTypes2.default.bool,

    /** Additional props are spread to the container component */
    '': _propTypes2.default.any // eslint-disable-line react/require-default-props

  };
  const defaultProps = {
    children: null,
    className: '',
    isOpen: true,
    openedOnEnter: false,
    position: _dropdownConstants.DropdownPosition.left,
    component: 'ul',
    isGrouped: false
  };

  class DropdownMenu extends _react2.default.Component {
    constructor(...args) {
      super(...args);

      _defineProperty(this, "refsCollection", []);

      _defineProperty(this, "childKeyHandler", (index, position, custom = false) => {
        (0, _util.keyHandler)(index, position, this.refsCollection, this.props.isGrouped ? this.refsCollection : this.props.children, custom);
      });

      _defineProperty(this, "sendRef", (index, node, isDisabled) => {
        if (!node.getAttribute) {
          this.refsCollection[index] = _reactDom2.default.findDOMNode(node); // eslint-disable-line react/no-find-dom-node
        } else if (isDisabled || node.getAttribute('role') === 'separator') {
          this.refsCollection[index] = null;
        } else {
          this.refsCollection[index] = node;
        }
      });
    }

    componentDidMount() {
      const focusTarget = this.refsCollection.filter(ref => ref && (ref.current && !ref.current.hasAttribute('disabled') || !ref.hasAttribute('disabled')))[0] || null;
      if (this.props.component === 'ul') focusTarget && focusTarget.focus();else {
        focusTarget.current.focus && focusTarget.current.focus() || focusTarget && _reactDom2.default.findDOMNode(focusTarget.current).focus(); // eslint-disable-line react/no-find-dom-node
      }
    }

    extendChildren() {
      const {
        children,
        isGrouped
      } = this.props;

      if (isGrouped) {
        let index = 0;
        return _react2.default.Children.map(children, group => _react2.default.cloneElement(group, {
          children: group.props.children.constructor === Array && group.props.children.map(option => _react2.default.cloneElement(option, {
            index: index++
          })) || _react2.default.cloneElement(group.props.children, {
            index: index++
          })
        }));
      }

      return _react2.default.Children.map(children, (child, index) => _react2.default.cloneElement(child, {
        index
      }));
    }

    extendCustomChildren() {
      const mappedChildren = _react2.default.Children.map(this.props.children, (child, index) => {
        const mappedChild = _react2.default.cloneElement(child, {
          ref: _react2.default.createRef(),
          className: `${(0, _reactStyles.css)(child.props.isDisabled && _dropdown2.default.modifiers.disabled, child.props.isHovered && _dropdown2.default.modifiers.hover, _dropdown2.default.dropdownMenuItem)}${child.props.className ? child.props.className : ''}`,
          tabIndex: -1,
          onKeyDown: event => {
            if (event.keyCode === _constants.KEY_CODES.TAB) return;
            event.preventDefault();

            if (event.keyCode === _constants.KEY_CODES.ARROW_UP) {
              (0, _util.keyHandler)(index, _constants.KEYHANDLER_DIRECTION.up, this.refsCollection, this.props.children, true);
            } else if (event.keyCode === _constants.KEY_CODES.ARROW_DOWN) {
              (0, _util.keyHandler)(index, _constants.KEYHANDLER_DIRECTION.DOWN, this.refsCollection, this.props.children, true);
            }
          }
        });

        !mappedChild.props.disabled ? this.refsCollection[index] = mappedChild.ref : this.refsCollection[index] = null;
        return mappedChild;
      });

      return mappedChildren;
    }

    render() {
      const _this$props = this.props,
            {
        className,
        isOpen,
        position,
        children,
        component: Component,
        isGrouped,
        openedOnEnter
      } = _this$props,
            props = _objectWithoutProperties(_this$props, ["className", "isOpen", "position", "children", "component", "isGrouped", "openedOnEnter"]);

      return _react2.default.createElement(_dropdownConstants.DropdownArrowContext.Provider, {
        value: {
          keyHandler: this.childKeyHandler,
          sendRef: this.sendRef
        }
      }, Component === 'div' ? _react2.default.createElement(_dropdownConstants.DropdownContext.Consumer, null, ({
        onSelect,
        menuClass
      }) => _react2.default.createElement("ul", {
        className: (0, _reactStyles.css)(menuClass, position === _dropdownConstants.DropdownPosition.right && _dropdown2.default.modifiers.alignRight, className),
        hidden: !isOpen,
        onClick: event => onSelect && onSelect(event)
      }, _react2.default.createElement(Component, props, this.extendCustomChildren()))) : isGrouped && _react2.default.createElement(_dropdownConstants.DropdownContext.Consumer, null, ({
        menuClass
      }) => _react2.default.createElement("div", _extends({}, props, {
        className: (0, _reactStyles.css)(menuClass, position === _dropdownConstants.DropdownPosition.right && _dropdown2.default.modifiers.alignRight, className),
        hidden: !isOpen,
        role: "menu"
      }), this.extendChildren())) || _react2.default.createElement(_dropdownConstants.DropdownContext.Consumer, null, ({
        menuClass
      }) => _react2.default.createElement(Component, _extends({}, props, {
        className: (0, _reactStyles.css)(menuClass, position === _dropdownConstants.DropdownPosition.right && _dropdown2.default.modifiers.alignRight, className),
        hidden: !isOpen,
        role: "menu"
      }), this.extendChildren())));
    }

  }

  DropdownMenu.propTypes = propTypes;
  DropdownMenu.defaultProps = defaultProps;
  exports.default = DropdownMenu;
});
