// @patternfly/react-core/dist/umd/components/Dropdown/DropdownItem.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "./InternalDropdownItem", "../../helpers/componentShape", "./dropdownConstants"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("./InternalDropdownItem"), require("../../helpers/componentShape"), require("./dropdownConstants"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.InternalDropdownItem, global.componentShape, global.dropdownConstants);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _InternalDropdownItem, _componentShape, _dropdownConstants) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _InternalDropdownItem2 = _interopRequireDefault(_InternalDropdownItem);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  const DropdownItem = _ref => {
    let props = _extends({}, _ref);

    return _react2.default.createElement(_dropdownConstants.DropdownArrowContext.Consumer, null, context => _react2.default.createElement(_InternalDropdownItem2.default, _extends({
      context: context,
      role: "menuitem",
      tabIndex: -1
    }, props)));
  };

  DropdownItem.propTypes = {
    /** Anything which can be rendered as dropdown item */
    children: _propTypes2.default.node,

    /** Classes applied to root element of dropdown item */
    className: _propTypes2.default.string,

    /** Indicates which component will be used as dropdown item */
    component: _componentShape.componentShape,

    /** Render dropdown item as disabled option */
    isDisabled: _propTypes2.default.bool,

    /** Forces display of the hover state of the element */
    isHovered: _propTypes2.default.bool,

    /** Default hyperlink location */
    href: _propTypes2.default.string,

    /** Tooltip to display when hovered over the item */
    tooltip: _propTypes2.default.node,

    /** Additional tooltip props forwarded to the Tooltip component */
    tooltipProps: _propTypes2.default.any,

    /** Additional props are passed to the DropdownItem */
    '': _propTypes2.default.any // eslint-disable-line react/require-default-props

  };
  DropdownItem.defaultProps = {
    children: null,
    className: '',
    isHovered: false,
    component: 'a',
    isDisabled: false,
    href: '',
    tooltip: null,
    tooltipProps: {}
  };
  exports.default = DropdownItem;
});
