// @patternfly/react-core/dist/umd/components/Dropdown/Toggle.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "@patternfly/react-styles/css/components/Dropdown/dropdown", "./dropdownConstants", "@patternfly/react-styles", "../../helpers/constants"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("@patternfly/react-styles/css/components/Dropdown/dropdown"), require("./dropdownConstants"), require("@patternfly/react-styles"), require("../../helpers/constants"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global.dropdown, global.dropdownConstants, global.reactStyles, global.constants);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _dropdown, _dropdownConstants, _reactStyles, _constants) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _dropdown2 = _interopRequireDefault(_dropdown);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  const propTypes = {
    /** HTML ID of dropdown toggle */
    id: _propTypes2.default.string.isRequired,

    /** Type to put on the button */
    type: _propTypes2.default.string,

    /** Anything which can be rendered as dropdown toggle */
    children: _propTypes2.default.node,

    /** Classes applied to root element of dropdown toggle */
    className: _propTypes2.default.string,

    /** Flag to indicate if menu is opened */
    isOpen: _propTypes2.default.bool,

    /** Callback called when toggle is clicked */
    onToggle: _propTypes2.default.func,

    /** Callback called when the Enter key is pressed */
    onEnter: _propTypes2.default.func,

    /** Element which wraps toggle */
    parentRef: _propTypes2.default.any,

    /** Forces focus state */
    isFocused: _propTypes2.default.bool,

    /** Forces hover state */
    isHovered: _propTypes2.default.bool,

    /** Forces active state */
    isActive: _propTypes2.default.bool,

    /** Disables the dropdown toggle */
    isDisabled: _propTypes2.default.bool,

    /** Display the toggle with no border or background */
    isPlain: _propTypes2.default.bool,

    /** Style the toggle as a child of a split button */
    isSplitButton: _propTypes2.default.bool,

    /** Flag for aria popup */
    ariaHasPopup: _propTypes2.default.bool,

    /** Additional props are spread to the container <button> */
    '': _propTypes2.default.any // eslint-disable-line react/require-default-props

  };
  const defaultProps = {
    children: null,
    className: '',
    type: null,
    isOpen: false,
    parentRef: null,
    isFocused: false,
    isHovered: false,
    isActive: false,
    isDisabled: false,
    isPlain: false,
    isSplitButton: false,
    ariaHasPopup: undefined,
    onToggle: Function.prototype,
    onEnter: Function.prototype
  };

  class Toggle extends _react.Component {
    constructor(...args) {
      super(...args);

      _defineProperty(this, "componentDidMount", () => {
        document.addEventListener('mousedown', this.onDocClick);
        document.addEventListener('touchstart', this.onDocClick);
        document.addEventListener('keydown', this.onEscPress);
      });

      _defineProperty(this, "componentWillUnmount", () => {
        document.removeEventListener('mousedown', this.onDocClick);
        document.removeEventListener('touchstart', this.onDocClick);
        document.removeEventListener('keydown', this.onEscPress);
      });

      _defineProperty(this, "onDocClick", event => {
        if (this.props.isOpen && this.props.parentRef && !this.props.parentRef.contains(event.target)) {
          this.props.onToggle && this.props.onToggle(false, event);
          this.toggle.focus();
        }
      });

      _defineProperty(this, "onEscPress", event => {
        const {
          parentRef
        } = this.props;
        const keyCode = event.keyCode || event.which;

        if (this.props.isOpen && (keyCode === _constants.KEY_CODES.ESCAPE_KEY || event.key === 'Tab') && parentRef && parentRef.contains(event.target)) {
          this.props.onToggle && this.props.onToggle(false, event);
          this.toggle.focus();
        }
      });

      _defineProperty(this, "onKeyDown", event => {
        if (event.key === 'Tab' && !this.props.isOpen) return;
        event.preventDefault();

        if ((event.key === 'Tab' || event.key === 'Enter' || event.key === ' ') && this.props.isOpen) {
          this.props.onToggle(!this.props.isOpen, event);
        } else if ((event.key === 'Enter' || event.key === ' ') && !this.props.isOpen) {
          this.props.onToggle(!this.props.isOpen, event);
          this.props.onEnter();
        }
      });
    }

    render() {
      const _this$props = this.props,
            {
        className,
        children,
        isOpen,
        isFocused,
        isActive,
        isHovered,
        isDisabled,
        isPlain,
        isSplitButton,
        ariaHasPopup,
        onToggle,
        onEnter,
        parentRef,
        id,
        type
      } = _this$props,
            props = _objectWithoutProperties(_this$props, ["className", "children", "isOpen", "isFocused", "isActive", "isHovered", "isDisabled", "isPlain", "isSplitButton", "ariaHasPopup", "onToggle", "onEnter", "parentRef", "id", "type"]);

      return _react2.default.createElement(_dropdownConstants.DropdownContext.Consumer, null, ({
        toggleClass
      }) => _react2.default.createElement("button", _extends({}, props, {
        id: id,
        ref: toggle => {
          this.toggle = toggle;
        },
        className: (0, _reactStyles.css)(isSplitButton ? _dropdown2.default.dropdownToggleButton : toggleClass || _dropdown2.default.dropdownToggle, isFocused && _dropdown2.default.modifiers.focus, isHovered && _dropdown2.default.modifiers.hover, isActive && _dropdown2.default.modifiers.active, isPlain && _dropdown2.default.modifiers.plain, isDisabled && _dropdown2.default.modifiers.disabled, className),
        type: type || 'button',
        onClick: event => onToggle && onToggle(!isOpen, event),
        "aria-expanded": isOpen,
        "aria-haspopup": ariaHasPopup,
        onKeyDown: this.onKeyDown,
        disabled: isDisabled
      }), children));
    }

  }

  Toggle.propTypes = propTypes;
  Toggle.defaultProps = defaultProps;
  exports.default = Toggle;
});
