// @patternfly/react-table/dist/js/test-helpers/data-sets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.actions = exports.rows = exports.columns = void 0;

var _react = _interopRequireDefault(require("react"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/* eslint-disable no-console */
var columns = [{
  title: 'Header cell'
}, 'Branches', {
  title: 'Pull requests'
}, 'Workspaces', {
  title: 'Last Commit'
}];
exports.columns = columns;
var rows = [{
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}, {
  cells: ['one', 'two', 'three', 'four', 'five']
}];
exports.rows = rows;
var actions = [{
  title: 'Some action',
  onClick: function onClick(event, rowId) {
    return console.log('clicked on Some action, on row: ', rowId);
  }
}, {
  title: _react["default"].createElement("div", null, "Another action"),
  onClick: function onClick(event, rowId) {
    return console.log('clicked on Another action, on row: ', rowId);
  }
}, {
  isSeparator: true
}, {
  title: 'Third action',
  onClick: function onClick(event, rowId) {
    return console.log('clicked on Third action, on row: ', rowId);
  }
}];
exports.actions = actions;
