// @patternfly/react-table/dist/esm/components/Table/ActionsColumn.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown, KebabToggle, DropdownItem, DropdownSeparator, DropdownPosition, DropdownDirection } from '@patternfly/react-core';
const propTypes = {
  children: PropTypes.node,
  items: PropTypes.array,
  isDisabled: PropTypes.bool,
  dropdownPosition: PropTypes.oneOf(Object.values(DropdownPosition)),
  dropdownDirection: PropTypes.oneOf(Object.values(DropdownDirection)),
  rowData: PropTypes.object,
  extraData: PropTypes.shape({
    rowIndex: PropTypes.number,
    columnIndex: PropTypes.number,
    column: PropTypes.object,
    property: PropTypes.string
  })
};
const defaultProps = {
  children: null,
  items: [],
  isDisabled: false,
  dropdownPosition: DropdownPosition.right,
  dropdownDirection: DropdownDirection.down,
  rowData: {},
  extraData: {}
};

class ActionsColumn extends React.Component {
  constructor(props) {
    super(props);

    _defineProperty(this, "onToggle", isOpen => {
      this.setState({
        isOpen
      });
    });

    _defineProperty(this, "onSelect", (event, onClick) => {
      const {
        rowData,
        extraData
      } = this.props;
      event.preventDefault();
      onClick && onClick(event, extraData.rowIndex, rowData, extraData);
      this.setState({
        isOpen: !this.state.isOpen
      });
    });

    this.state = {
      isOpen: false
    };
  }

  render() {
    const {
      isOpen
    } = this.state;
    const {
      items,
      children,
      dropdownPosition,
      dropdownDirection,
      isDisabled
    } = this.props;
    return React.createElement(React.Fragment, null, React.createElement(Dropdown, {
      onToggle: this.onToggle,
      toggle: React.createElement(KebabToggle, {
        isDisabled: isDisabled,
        onToggle: this.onToggle
      }),
      position: dropdownPosition,
      direction: dropdownDirection,
      isOpen: isOpen,
      dropdownItems: items.map((_ref, key) => {
        let {
          title,
          itemKey,
          onClick,
          isSeparator
        } = _ref,
            props = _objectWithoutProperties(_ref, ["title", "itemKey", "onClick", "isSeparator"]);

        return isSeparator ? React.createElement(DropdownSeparator, _extends({}, props, {
          key: itemKey || key,
          "data-key": itemKey || key
        })) : React.createElement(DropdownItem, _extends({
          onClick: event => this.onSelect(event, onClick)
        }, props, {
          key: itemKey || key,
          "data-key": itemKey || key
        }), title);
      }),
      isPlain: true
    }), children);
  }

}

ActionsColumn.propTypes = propTypes;
ActionsColumn.defaultProps = defaultProps;
export default ActionsColumn;
