// @patternfly/react-core/dist/umd/layouts/Grid/gridUtils.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "@patternfly/react-styles/css/layouts/Grid/grid", "@patternfly/react-styles"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("@patternfly/react-styles/css/layouts/Grid/grid"), require("@patternfly/react-styles"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.grid, global.reactStyles);
    global.undefined = mod.exports;
  }
})(this, function (exports, _grid, _reactStyles) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.getOffsetModifier = exports.getRowSpanModifier = exports.getGridSpanModifier = exports.getSpanModifier = exports.getSizeSuffix = exports.getOffsetKey = exports.getRowSpanKey = exports.gridSpans = undefined;

  var _grid2 = _interopRequireDefault(_grid);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const gridSpans = exports.gridSpans = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];

  const getRowSpanKey = exports.getRowSpanKey = size => `${size}RowSpan`;

  const getOffsetKey = exports.getOffsetKey = size => `${size}Offset`;

  const getSizeSuffix = exports.getSizeSuffix = size => size ? `-on-${size}` : '';

  const getSpanModifier = exports.getSpanModifier = (value, size) => (0, _reactStyles.getModifier)(_grid2.default, `${value}-col${getSizeSuffix(size)}`);

  const getGridSpanModifier = exports.getGridSpanModifier = (value, size) => (0, _reactStyles.getModifier)(_grid2.default, `-all-${value}-col${getSizeSuffix(size)}`);

  const getRowSpanModifier = exports.getRowSpanModifier = (value, size) => (0, _reactStyles.getModifier)(_grid2.default, `${value}-row${getSizeSuffix(size)}`);

  const getOffsetModifier = exports.getOffsetModifier = (value, size) => (0, _reactStyles.getModifier)(_grid2.default, `offset-${value}-col${getSizeSuffix(size)}`);
});
