// @patternfly/react-core/dist/esm/components/Tooltip/Tooltip.js
import _pt from "prop-types";

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

import * as React from 'react';
import Tippy from '@tippy.js/react';
import styles from '@patternfly/react-styles/css/components/Tooltip/tooltip';
import { css, getModifier } from '@patternfly/react-styles';
import { TooltipArrow } from './TooltipArrow';
import { TooltipContent } from './TooltipContent';
import { KEY_CODES } from '../../helpers/constants';
import { c_tooltip_MaxWidth as tooltipMaxWidth } from '@patternfly/react-tokens';
import { tippyStyles } from './styles';
tippyStyles();
export let TooltipPosition;

(function (TooltipPosition) {
  TooltipPosition["top"] = "top";
  TooltipPosition["bottom"] = "bottom";
  TooltipPosition["left"] = "left";
  TooltipPosition["right"] = "right";
})(TooltipPosition || (TooltipPosition = {}));

;
;
export class Tooltip extends React.Component {
  constructor(...args) {
    super(...args);

    _defineProperty(this, "storeTippyInstance", tip => {
      this.tip = tip;
    });

    _defineProperty(this, "handleEscKeyClick", event => {
      if (event.keyCode === KEY_CODES.ESCAPE_KEY && this.tip.state.isVisible) {
        this.tip.hide();
      }
    });
  }

  componentDidMount() {
    document.addEventListener('keydown', this.handleEscKeyClick, false);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.handleEscKeyClick, false);
  }

  extendChildren() {
    return React.cloneElement(this.props.children, {
      isAppLauncher: this.props.isAppLauncher
    });
  }

  render() {
    const _this$props = this.props,
          {
      position,
      trigger,
      enableFlip,
      children,
      className,
      content: bodyContent,
      entryDelay,
      exitDelay,
      appendTo,
      zIndex,
      maxWidth,
      isAppLauncher,
      distance
    } = _this$props,
          rest = _objectWithoutProperties(_this$props, ["position", "trigger", "enableFlip", "children", "className", "content", "entryDelay", "exitDelay", "appendTo", "zIndex", "maxWidth", "isAppLauncher", "distance"]);

    const content = React.createElement("div", _extends({
      className: css(styles.tooltip, !enableFlip && getModifier(styles, position, styles.modifiers.top), className),
      role: "tooltip"
    }, rest), React.createElement(TooltipArrow, null), React.createElement(TooltipContent, null, bodyContent));
    return React.createElement(Tippy, {
      onCreate: this.storeTippyInstance,
      maxWidth: maxWidth,
      zIndex: zIndex,
      appendTo: appendTo,
      content: content,
      lazy: true,
      animateFill: false,
      theme: "pf-tippy",
      performance: true,
      placement: position,
      trigger: trigger,
      delay: [entryDelay, exitDelay],
      distance: distance,
      flip: enableFlip,
      popperOptions: {
        modifiers: {
          preventOverflow: {
            enabled: enableFlip
          },
          hide: {
            enabled: enableFlip
          }
        }
      }
    }, isAppLauncher ? this.extendChildren() : children);
  }

}

_defineProperty(Tooltip, "propTypes", {
  position: _pt.oneOf(['top', 'bottom', 'left', 'right']),
  trigger: _pt.string,
  enableFlip: _pt.bool,
  className: _pt.string,
  content: _pt.node.isRequired,
  children: _pt.element.isRequired,
  entryDelay: _pt.number,
  exitDelay: _pt.number,
  appendTo: _pt.oneOfType([_pt.element, _pt.func]),
  zIndex: _pt.number,
  maxWidth: _pt.string,
  isAppLauncher: _pt.bool,
  distance: _pt.number
});

_defineProperty(Tooltip, "defaultProps", {
  position: 'top',
  trigger: 'mouseenter focus',
  enableFlip: true,
  className: '',
  entryDelay: 500,
  exitDelay: 500,
  appendTo: () => document.body,
  zIndex: 9999,
  maxWidth: tooltipMaxWidth && tooltipMaxWidth.value,
  isAppLauncher: false,
  distance: 15
});
