// @patternfly/react-table/dist/js/components/Table/SortColumn.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = exports.SortByDirection = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactIcons = require("@patternfly/react-icons");

var _reactStyles = require("@patternfly/react-styles");

var _table = _interopRequireDefault(require("@patternfly/react-styles/css/components/Table/table"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

var propTypes = {
  children: _propTypes["default"].node,
  className: _propTypes["default"].string,
  isSortedBy: _propTypes["default"].bool,
  onSort: _propTypes["default"].func,
  sortDirection: _propTypes["default"].string
};
var defaultProps = {
  children: null,
  className: '',
  isSortedBy: false,
  onSort: null,
  sortDirection: ''
};
var SortByDirection = {
  asc: 'asc',
  desc: 'desc'
};
exports.SortByDirection = SortByDirection;

var SortColumn = function SortColumn(_ref) {
  var isSortedBy = _ref.isSortedBy,
      children = _ref.children,
      className = _ref.className,
      onSort = _ref.onSort,
      sortDirection = _ref.sortDirection,
      props = _objectWithoutProperties(_ref, ["isSortedBy", "children", "className", "onSort", "sortDirection"]);

  var SortedByIcon;

  if (isSortedBy) {
    SortedByIcon = sortDirection === 'asc' ? _reactIcons.LongArrowAltUpIcon : _reactIcons.LongArrowAltDownIcon;
  } else {
    SortedByIcon = _reactIcons.ArrowsAltVIcon;
  }

  return _react["default"].createElement("button", _extends({}, props, {
    className: (0, _reactStyles.css)(className),
    onClick: function onClick(event) {
      return onSort && onSort(event);
    }
  }), children, _react["default"].createElement("span", {
    className: (0, _reactStyles.css)(_table["default"].tableSortIndicator)
  }, _react["default"].createElement(SortedByIcon, null)));
};

SortColumn.propTypes = propTypes;
SortColumn.defaultProps = defaultProps;
var _default = SortColumn;
exports["default"] = _default;
