// @patternfly/react-core/dist/umd/components/Tooltip/Tooltip.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "prop-types", "react", "@tippy.js/react", "@patternfly/react-styles/css/components/Tooltip/tooltip", "@patternfly/react-styles", "./TooltipArrow", "./TooltipContent", "../../helpers/constants", "@patternfly/react-tokens", "./styles"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("prop-types"), require("react"), require("@tippy.js/react"), require("@patternfly/react-styles/css/components/Tooltip/tooltip"), require("@patternfly/react-styles"), require("./TooltipArrow"), require("./TooltipContent"), require("../../helpers/constants"), require("@patternfly/react-tokens"), require("./styles"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.propTypes, global.react, global.react, global.tooltip, global.reactStyles, global.TooltipArrow, global.TooltipContent, global.constants, global.reactTokens, global.styles);
    global.undefined = mod.exports;
  }
})(this, function (exports, _propTypes, _react, _react2, _tooltip, _reactStyles, _TooltipArrow, _TooltipContent, _constants, _reactTokens, _styles) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.Tooltip = exports.TooltipPosition = undefined;

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var React = _interopRequireWildcard(_react);

  var _react3 = _interopRequireDefault(_react2);

  var _tooltip2 = _interopRequireDefault(_tooltip);

  function _interopRequireWildcard(obj) {
    if (obj && obj.__esModule) {
      return obj;
    } else {
      var newObj = {};

      if (obj != null) {
        for (var key in obj) {
          if (Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {};

            if (desc.get || desc.set) {
              Object.defineProperty(newObj, key, desc);
            } else {
              newObj[key] = obj[key];
            }
          }
        }
      }

      newObj.default = obj;
      return newObj;
    }
  }

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  (0, _styles.tippyStyles)();
  let TooltipPosition = exports.TooltipPosition = undefined;

  (function (TooltipPosition) {
    TooltipPosition["top"] = "top";
    TooltipPosition["bottom"] = "bottom";
    TooltipPosition["left"] = "left";
    TooltipPosition["right"] = "right";
  })(TooltipPosition || (exports.TooltipPosition = TooltipPosition = {}));

  ;
  ;

  class Tooltip extends React.Component {
    constructor(...args) {
      super(...args);

      _defineProperty(this, "storeTippyInstance", tip => {
        this.tip = tip;
      });

      _defineProperty(this, "handleEscKeyClick", event => {
        if (event.keyCode === _constants.KEY_CODES.ESCAPE_KEY && this.tip.state.isVisible) {
          this.tip.hide();
        }
      });
    }

    componentDidMount() {
      document.addEventListener('keydown', this.handleEscKeyClick, false);
    }

    componentWillUnmount() {
      document.removeEventListener('keydown', this.handleEscKeyClick, false);
    }

    extendChildren() {
      return React.cloneElement(this.props.children, {
        isAppLauncher: this.props.isAppLauncher
      });
    }

    render() {
      const _this$props = this.props,
            {
        position,
        trigger,
        enableFlip,
        children,
        className,
        content: bodyContent,
        entryDelay,
        exitDelay,
        appendTo,
        zIndex,
        maxWidth,
        isAppLauncher,
        distance
      } = _this$props,
            rest = _objectWithoutProperties(_this$props, ["position", "trigger", "enableFlip", "children", "className", "content", "entryDelay", "exitDelay", "appendTo", "zIndex", "maxWidth", "isAppLauncher", "distance"]);

      const content = React.createElement("div", _extends({
        className: (0, _reactStyles.css)(_tooltip2.default.tooltip, !enableFlip && (0, _reactStyles.getModifier)(_tooltip2.default, position, _tooltip2.default.modifiers.top), className),
        role: "tooltip"
      }, rest), React.createElement(_TooltipArrow.TooltipArrow, null), React.createElement(_TooltipContent.TooltipContent, null, bodyContent));
      return React.createElement(_react3.default, {
        onCreate: this.storeTippyInstance,
        maxWidth: maxWidth,
        zIndex: zIndex,
        appendTo: appendTo,
        content: content,
        lazy: true,
        animateFill: false,
        theme: "pf-tippy",
        performance: true,
        placement: position,
        trigger: trigger,
        delay: [entryDelay, exitDelay],
        distance: distance,
        flip: enableFlip,
        popperOptions: {
          modifiers: {
            preventOverflow: {
              enabled: enableFlip
            },
            hide: {
              enabled: enableFlip
            }
          }
        }
      }, isAppLauncher ? this.extendChildren() : children);
    }

  }

  exports.Tooltip = Tooltip;

  _defineProperty(Tooltip, "propTypes", {
    position: _propTypes2.default.oneOf(['top', 'bottom', 'left', 'right']),
    trigger: _propTypes2.default.string,
    enableFlip: _propTypes2.default.bool,
    className: _propTypes2.default.string,
    content: _propTypes2.default.node.isRequired,
    children: _propTypes2.default.element.isRequired,
    entryDelay: _propTypes2.default.number,
    exitDelay: _propTypes2.default.number,
    appendTo: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.func]),
    zIndex: _propTypes2.default.number,
    maxWidth: _propTypes2.default.string,
    isAppLauncher: _propTypes2.default.bool,
    distance: _propTypes2.default.number
  });

  _defineProperty(Tooltip, "defaultProps", {
    position: 'top',
    trigger: 'mouseenter focus',
    enableFlip: true,
    className: '',
    entryDelay: 500,
    exitDelay: 500,
    appendTo: () => document.body,
    zIndex: 9999,
    maxWidth: _reactTokens.c_tooltip_MaxWidth && _reactTokens.c_tooltip_MaxWidth.value,
    isAppLauncher: false,
    distance: 15
  });
});
