// @patternfly/react-core/dist/umd/components/Switch/Switch.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "prop-types", "@patternfly/react-styles/css/components/Switch/switch", "@patternfly/react-styles", "@patternfly/react-icons", "../../helpers/util"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("prop-types"), require("@patternfly/react-styles/css/components/Switch/switch"), require("@patternfly/react-styles"), require("@patternfly/react-icons"), require("../../helpers/util"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.propTypes, global._switch, global.reactStyles, global.reactIcons, global.util);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _propTypes, _switch, _reactStyles, _reactIcons, _util) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _propTypes2 = _interopRequireDefault(_propTypes);

  var _switch2 = _interopRequireDefault(_switch);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  function _extends() {
    _extends = Object.assign || function (target) {
      for (var i = 1; i < arguments.length; i++) {
        var source = arguments[i];

        for (var key in source) {
          if (Object.prototype.hasOwnProperty.call(source, key)) {
            target[key] = source[key];
          }
        }
      }

      return target;
    };

    return _extends.apply(this, arguments);
  }

  function _objectWithoutProperties(source, excluded) {
    if (source == null) return {};

    var target = _objectWithoutPropertiesLoose(source, excluded);

    var key, i;

    if (Object.getOwnPropertySymbols) {
      var sourceSymbolKeys = Object.getOwnPropertySymbols(source);

      for (i = 0; i < sourceSymbolKeys.length; i++) {
        key = sourceSymbolKeys[i];
        if (excluded.indexOf(key) >= 0) continue;
        if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue;
        target[key] = source[key];
      }
    }

    return target;
  }

  function _objectWithoutPropertiesLoose(source, excluded) {
    if (source == null) return {};
    var target = {};
    var sourceKeys = Object.keys(source);
    var key, i;

    for (i = 0; i < sourceKeys.length; i++) {
      key = sourceKeys[i];
      if (excluded.indexOf(key) >= 0) continue;
      target[key] = source[key];
    }

    return target;
  }

  function _defineProperty(obj, key, value) {
    if (key in obj) {
      Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
      });
    } else {
      obj[key] = value;
    }

    return obj;
  }

  const propTypes = {
    /** id for the label. */
    id: _propTypes2.default.string,

    /** Additional classes added to the Switch */
    className: _propTypes2.default.string,

    /** Text value for the label */
    label: _propTypes2.default.string,

    /** Flag to show if the Switch is checked. */
    isChecked: _propTypes2.default.bool,

    /** Flag to show if the Switch is disabled. */
    isDisabled: _propTypes2.default.bool,

    /** A callback for when the Switch selection changes. (isChecked, event) => {} */
    onChange: _propTypes2.default.func,

    /** Adds accessible text to the Switch, and should describe the isChecked="true" state. When label is defined, aria-label should be set to the text string that is visible when isChecked is true. */
    'aria-label': props => {
      if (!props.id && !props['aria-label']) {
        return new Error('Switch requires either an id or aria-label to be specified');
      }

      return null;
    },

    /** Additional props are spread to the container <input> */
    '': _propTypes2.default.any // eslint-disable-line react/require-default-props

  };
  const defaultProps = {
    id: '',
    className: '',
    label: '',
    isChecked: true,
    isDisabled: false,
    onChange: () => undefined,
    'aria-label': ''
  };

  class Switch extends _react2.default.Component {
    constructor(...args) {
      super(...args);

      _defineProperty(this, "id", this.props.id || (0, _util.getUniqueId)());
    }

    render() {
      const _this$props = this.props,
            {
        id,
        className,
        label,
        isChecked,
        isDisabled,
        onChange
      } = _this$props,
            props = _objectWithoutProperties(_this$props, ["id", "className", "label", "isChecked", "isDisabled", "onChange"]);

      return _react2.default.createElement("label", {
        className: (0, _reactStyles.css)(_switch2.default.switch, className),
        htmlFor: this.id
      }, _react2.default.createElement("input", _extends({}, props, {
        id: this.id,
        className: (0, _reactStyles.css)(_switch2.default.switchInput),
        type: "checkbox",
        onChange: event => onChange(event.currentTarget.checked, event),
        checked: isChecked,
        disabled: isDisabled
      })), label !== '' ? _react2.default.createElement(_react2.default.Fragment, null, _react2.default.createElement("span", {
        className: (0, _reactStyles.css)(_switch2.default.switchToggle)
      }), _react2.default.createElement("span", {
        className: (0, _reactStyles.css)(_switch2.default.switchLabel, _switch2.default.modifiers.on),
        "aria-hidden": "true"
      }, label), _react2.default.createElement("span", {
        className: (0, _reactStyles.css)(_switch2.default.switchLabel, _switch2.default.modifiers.off),
        "aria-hidden": "true"
      }, label)) : _react2.default.createElement("span", {
        className: (0, _reactStyles.css)(_switch2.default.switchToggle)
      }, _react2.default.createElement("div", {
        className: (0, _reactStyles.css)(_switch2.default.switchToggleIcon),
        "aria-hidden": "true"
      }, _react2.default.createElement(_reactIcons.CheckIcon, {
        noVerticalAlign: true
      }))));
    }

  }

  Switch.propTypes = propTypes;
  Switch.defaultProps = defaultProps;
  exports.default = Switch;
});
