// @patternfly/react-table/dist/umd/components/Table/utils/decorators/collapsible.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "@patternfly/react-styles", "@patternfly/react-styles/css/components/Table/table", "../../CollapseColumn", "../../ExpandableRowContent"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("@patternfly/react-styles"), require("@patternfly/react-styles/css/components/Table/table"), require("../../CollapseColumn"), require("../../ExpandableRowContent"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactStyles, global.table, global.CollapseColumn, global.ExpandableRowContent);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _reactStyles, _table, _CollapseColumn, _ExpandableRowContent) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.expandedRow = exports.expandable = exports.collapsible = undefined;

  var _react2 = _interopRequireDefault(_react);

  var _table2 = _interopRequireDefault(_table);

  var _CollapseColumn2 = _interopRequireDefault(_CollapseColumn);

  var _ExpandableRowContent2 = _interopRequireDefault(_ExpandableRowContent);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const collapsible = exports.collapsible = (value, {
    rowIndex,
    columnIndex,
    rowData,
    column,
    property
  }) => {
    const {
      extraParams: {
        onCollapse,
        rowLabeledBy = 'simple-node',
        expandId = 'expand-toggle'
      }
    } = column;
    const extraData = {
      rowIndex,
      columnIndex,
      column,
      property
    };

    function onToggle(event) {
      onCollapse && onCollapse(event, rowIndex, rowData && !rowData.isOpen, rowData, extraData);
    }

    return {
      className: rowData.isOpen !== undefined && (0, _reactStyles.css)(_table2.default.tableToggle),
      isVisible: !rowData.fullWidth,
      children: _react2.default.createElement(_CollapseColumn2.default, {
        "aria-labelledby": `${rowLabeledBy}${rowIndex} ${expandId}${rowIndex}`,
        onToggle: onToggle,
        id: expandId + rowIndex,
        isOpen: rowData && rowData.isOpen
      }, value)
    };
  };

  const expandable = exports.expandable = (value, {
    rowData
  }) => rowData.hasOwnProperty('parent') ? _react2.default.createElement(_ExpandableRowContent2.default, null, value) : value;

  const expandedRow = exports.expandedRow = colSpan => {
    const expandedRowFormatter = (value, {
      rowIndex,
      rowData,
      column: {
        extraParams: {
          contentId = 'expanded-content'
        }
      }
    }) => rowData.hasOwnProperty('parent') && {
      colSpan: colSpan + !!rowData.fullWidth,
      id: contentId + rowIndex,
      className: rowData.noPadding && (0, _reactStyles.css)(_table2.default.modifiers.noPadding)
    };

    return expandedRowFormatter;
  };
});
