// @patternfly/react-table/dist/umd/components/Table/utils/decorators/compoundExpand.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "@patternfly/react-styles", "@patternfly/react-core", "@patternfly/react-styles/css/components/Table/table"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("@patternfly/react-styles"), require("@patternfly/react-core"), require("@patternfly/react-styles/css/components/Table/table"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.reactStyles, global.reactCore, global.table);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _reactStyles, _reactCore, _table) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });
  exports.compoundExpand = undefined;

  var _react2 = _interopRequireDefault(_react);

  var _table2 = _interopRequireDefault(_table);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  const compoundExpand = exports.compoundExpand = (value, {
    rowIndex,
    columnIndex,
    rowData,
    column,
    property
  }) => {
    if (!value) {
      return null;
    }

    const {
      title,
      props
    } = value;
    const {
      extraParams: {
        onExpand
      }
    } = column;
    const extraData = {
      rowIndex,
      columnIndex,
      column,
      property
    };

    function onToggle(event) {
      onExpand && onExpand(event, rowIndex, columnIndex, props.isOpen, rowData, extraData);
    }

    return {
      className: (0, _reactStyles.css)(_table2.default.tableCompoundExpansionToggle, props.isOpen && _table2.default.modifiers.expanded),
      children: props.isOpen !== undefined && _react2.default.createElement(_reactCore.Button, {
        variant: "link",
        onClick: onToggle,
        "aria-expanded": props.isOpen,
        "aria-controls": props.ariaControls
      }, title)
    };
  };
});
