// @patternfly/react-table/dist/umd/components/Table/utils/decorators/sortable.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "react", "../../SortColumn", "@patternfly/react-styles", "@patternfly/react-styles/css/components/Table/table", "@patternfly/react-styles/css/components/Button/button"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("react"), require("../../SortColumn"), require("@patternfly/react-styles"), require("@patternfly/react-styles/css/components/Table/table"), require("@patternfly/react-styles/css/components/Button/button"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.react, global.SortColumn, global.reactStyles, global.table, global.button);
    global.undefined = mod.exports;
  }
})(this, function (exports, _react, _SortColumn, _reactStyles, _table, _button) {
  "use strict";

  Object.defineProperty(exports, "__esModule", {
    value: true
  });

  var _react2 = _interopRequireDefault(_react);

  var _SortColumn2 = _interopRequireDefault(_SortColumn);

  var _table2 = _interopRequireDefault(_table);

  var _button2 = _interopRequireDefault(_button);

  function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
      default: obj
    };
  }

  exports.default = (label, {
    columnIndex,
    column,
    property
  }) => {
    const {
      extraParams: {
        sortBy,
        onSort
      }
    } = column;
    const extraData = {
      columnIndex,
      column,
      property
    };
    const isSortedBy = sortBy && columnIndex === sortBy.index;

    function sortClicked(event) {
      let reversedDirection;

      if (!isSortedBy) {
        reversedDirection = _SortColumn.SortByDirection.asc;
      } else {
        reversedDirection = sortBy.direction === _SortColumn.SortByDirection.asc ? _SortColumn.SortByDirection.desc : _SortColumn.SortByDirection.asc;
      }

      onSort && onSort(event, columnIndex, reversedDirection, extraData);
    }

    return {
      className: (0, _reactStyles.css)(_table2.default.tableSort, isSortedBy && _table2.default.modifiers.selected),
      'aria-sort': isSortedBy ? `${sortBy.direction}ending` : 'none',
      children: _react2.default.createElement(_SortColumn2.default, {
        isSortedBy: isSortedBy,
        sortDirection: isSortedBy ? sortBy.direction : '',
        onSort: sortClicked,
        className: (0, _reactStyles.css)(_button2.default.button, _button2.default.modifiers.plain)
      }, label)
    };
  };
});
