// @patternfly/react-core/dist/js/components/Dropdown/DropdownItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _InternalDropdownItem = _interopRequireDefault(require("./InternalDropdownItem"));

var _componentShape = require("../../helpers/componentShape");

var _dropdownConstants = require("./dropdownConstants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var DropdownItem = function DropdownItem(_ref) {
  var props = _extends({}, _ref);

  return _react["default"].createElement(_dropdownConstants.DropdownArrowContext.Consumer, null, function (context) {
    return _react["default"].createElement(_InternalDropdownItem["default"], _extends({
      context: context,
      role: "menuitem",
      tabIndex: -1
    }, props));
  });
};

DropdownItem.propTypes = {
  /** Anything which can be rendered as dropdown item */
  children: _propTypes["default"].node,

  /** Classes applied to root element of dropdown item */
  className: _propTypes["default"].string,

  /** Indicates which component will be used as dropdown item */
  component: _componentShape.componentShape,

  /** Render dropdown item as disabled option */
  isDisabled: _propTypes["default"].bool,

  /** Forces display of the hover state of the element */
  isHovered: _propTypes["default"].bool,

  /** Default hyperlink location */
  href: _propTypes["default"].string,

  /** Tooltip to display when hovered over the item */
  tooltip: _propTypes["default"].node,

  /** Additional tooltip props forwarded to the Tooltip component */
  tooltipProps: _propTypes["default"].any,

  /** Additional props are passed to the DropdownItem */
  '': _propTypes["default"].any // eslint-disable-line react/require-default-props

};
DropdownItem.defaultProps = {
  children: null,
  className: '',
  isHovered: false,
  component: 'a',
  isDisabled: false,
  href: '',
  tooltip: null,
  tooltipProps: {}
};
var _default = DropdownItem;
exports["default"] = _default;
